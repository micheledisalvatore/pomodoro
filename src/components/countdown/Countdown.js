import React from 'react';
import PropTypes from 'prop-types';

import {
  Container, Number, Svg, Circle,
} from './Countdown.styled';

export const Countdown = ({
  color,
  length,
  isPaused,
  timing,
}) => {
  const minutes = Math.floor(timing / 60).toString().padStart(2, '0');
  const seconds = (timing % 60).toString().padStart(2, '0');

  return (
    <Container>
      <Number color={color}>{minutes}:{seconds}</Number>
      { timing > 0 && (
        <Svg>
          <Circle duration={length} color={color} isPaused={isPaused} />
        </Svg>
      )}
    </Container>
  );
};

Countdown.propTypes = {
  color: PropTypes.string.isRequired,
  isPaused: PropTypes.bool.isRequired,
  length: PropTypes.number.isRequired,
  timing: PropTypes.number.isRequired,
};

export const mapStateToProps = ({
  session: {
    color,
    isPaused,
    length,
    timing,
  },
}) => ({
  color,
  length,
  isPaused,
  timing,
});
