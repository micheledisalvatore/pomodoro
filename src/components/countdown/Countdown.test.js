/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';

import { Number, Circle } from './Countdown.styled';
import { Countdown, mapStateToProps } from './Countdown';

describe('Given a Countdown component', () => {
  const COLOR_PROP = 'red';
  let component;
  let wrapper;

  const requiredProps = {
    color: COLOR_PROP,
    isPaused: true,
    length: 10,
    timing: 10,
  };

  describe('when it is rendered', () => {
    beforeEach(() => {
      component = <Countdown {...requiredProps} />;
      wrapper = shallow(component);
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should have the timer set to 10 sec', () => {
      expect(wrapper.find(Circle)).toHaveProp('duration', 10);
    });

    it(`should have the timer with the ${COLOR_PROP} color`, () => {
      expect(wrapper.find(Circle)).toHaveProp('color', COLOR_PROP);
    });

    it('should have the timer in pause', () => {
      expect(wrapper.find(Circle)).toHaveProp('isPaused', true);
    });

    it(`should show the countdown with the ${COLOR_PROP} color`, () => {
      expect(wrapper.find(Number)).toHaveProp('color', COLOR_PROP);
    });

    it('should show the countdown set to 00:10', () => {
      expect(wrapper.find(Number)).toHaveText('00:10');
    });

    describe('and the user starts the countdown', () => {
      beforeEach(() => {
        wrapper.setProps({
          isPaused: false,
        });
      });

      it('should match the snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should have the timer in play', () => {
        expect(wrapper.find(Circle)).toHaveProp('isPaused', false);
      });

      describe('and 3 seconds are passed', () => {
        beforeEach(() => {
          wrapper.setProps({
            timing: 7,
          });
        });

        it('should match the snapshot', () => {
          expect(wrapper).toMatchSnapshot();
        });

        it('should show the countdown set to 00:07', () => {
          expect(wrapper.find(Number)).toHaveText('00:07');
        });

        describe('and the timer ends', () => {
          beforeEach(() => {
            wrapper.setProps({
              timing: 0,
            });
          });

          it('should match the snapshot', () => {
            expect(wrapper).toMatchSnapshot();
          });

          it('should show the countdown set to 00:00', () => {
            expect(wrapper.find(Number)).toHaveText('00:00');
          });

          it('should NOT show the timer', () => {
            expect(wrapper.find(Circle)).not.toExist();
          });
        });
      });
    });
  });
});

describe('Given a mapStateToProps function', () => {
  describe('when it is called', () => {
    it('should return an object with the passed values', () => {
      expect(mapStateToProps({
        session: {
          color: 'fooColor',
          isPaused: 'fooIsPause',
          length: 'fooLength',
          timing: 'fooTiming',
        },
      })).toEqual({
        color: 'fooColor',
        isPaused: 'fooIsPause',
        length: 'fooLength',
        timing: 'fooTiming',
      });
    });
  });
});
