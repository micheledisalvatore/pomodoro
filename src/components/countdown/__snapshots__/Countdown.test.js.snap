// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Countdown component when it is rendered and the user starts the countdown and 3 seconds are passed and the timer ends should match the snapshot 1`] = `
<styled.div>
  <styled.div
    color="red"
  >
    00
    :
    00
  </styled.div>
</styled.div>
`;

exports[`Given a Countdown component when it is rendered and the user starts the countdown and 3 seconds are passed should match the snapshot 1`] = `
<styled.div>
  <styled.div
    color="red"
  >
    00
    :
    07
  </styled.div>
  <styled.svg>
    <styled.circle
      color="red"
      duration={10}
      isPaused={false}
    />
  </styled.svg>
</styled.div>
`;

exports[`Given a Countdown component when it is rendered and the user starts the countdown should match the snapshot 1`] = `
<styled.div>
  <styled.div
    color="red"
  >
    00
    :
    10
  </styled.div>
  <styled.svg>
    <styled.circle
      color="red"
      duration={10}
      isPaused={false}
    />
  </styled.svg>
</styled.div>
`;

exports[`Given a Countdown component when it is rendered should match the snapshot 1`] = `
<styled.div>
  <styled.div
    color="red"
  >
    00
    :
    10
  </styled.div>
  <styled.svg>
    <styled.circle
      color="red"
      duration={10}
      isPaused={true}
    />
  </styled.svg>
</styled.div>
`;
